confidant/services/certificatemanager.py

```diff
@@ -0,0 +1,449 @@
+        number of days), and a list of subject alternative names, return a dict
+        with the PEM encoded certificate, certificate chain, and private RSA
+        key.
+        """
+        with stats.timer('issue_certificate_with_key'):
+            cache_id = self.cache.get_cache_id(cn, validity, san)
+            cached_response = self._get_cached_certificate_with_key(cache_id)
+            if cached_response:
+                stats.incr('get_cached_certificate_with_key.hit')
+                logging.debug('Used cached response for {}'.format(cache_id))
+                return cached_response
+            stats.incr('get_cached_certificate_with_key.miss')
+            key = self.generate_key()
+            encoded_key = self.encode_key(key)
+            if self.settings['self_sign']:
+                cert = self.encode_certificate(
+                    self.generate_self_signed_certificate(
+                        key,
+                        cn,
+                        validity,
+                        san,
+                    )
+                )
+                return {
+                    'certificate': cert,
+                    'certificate_chain': cert,
+                    'key': encoded_key,
+                }
+            csr = self.generate_csr(key, cn, san)
+            try:
+                # set a lock
+                self.cache.lock(cache_id)
+                arn = self.issue_certificate(self.encode_csr(csr), validity)
+                response = self.get_certificate_from_arn(arn)
+                response['key'] = encoded_key
+                self.cache.set_response(cache_id, response)
+            finally:
+                # release the lock
+                self.cache.release(cache_id)
+            return response
+
+    def get_certificate_from_arn(self, certificate_arn):
+        """
+        Get the PEM encoded certificate from the provided ARN.
+        """
+        with stats.timer('get_certificate_from_arn'):
+            client = confidant.clients.get_boto_client('acm-pca')
+            # When a certificate is issued, it may take a while before it's
+            # available via get_certificate. We need to keep retrying until it's
+            # fully issued.
+            i = 0
+            while True:
+                try:
+                    response = client.get_certificate(
+                        CertificateAuthorityArn=self.settings['arn'],
+                        CertificateArn=certificate_arn,
+                    )
+                    break
+                except client.exceptions.RequestInProgressException:
+                    # Sleep for a maximum of 10 seconds
+                    if i >= 50:
+                        raise
+                    logging.debug(
+                        'Sleeping in get_certificate_from_arn for {}'.format(
+                            certificate_arn,
+                        )
+                    )
+                    time.sleep(.200)
```

Comment:
perhaps dumb q -- with a maximum sleep of 10seconds, would it be possible to DOS confidant if theoretically ACM goes down or becomes degraded, and a bunch of services request certs where the processes then become tied up?

Reply:
Hm. Yeah, that's definitely a possibility. It's a good concern. Maybe it's better to not sleep, but return a 429 here, which would cause the client to slow down and retry.

Reply:
Sending 429s with a Retry-After response header now